src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs

```diff
@@ -25,29 +24,73 @@ internal sealed partial class CSharpSelectionValidator(
+        protected override InitialSelectionInfo GetInitialSelectionInfo(CancellationToken cancellationToken)
+        {
+            var root = this.SemanticDocument.Root;
+
+            var adjustedSpan = GetAdjustedSpan(OriginalSpan);
+
+            var firstTokenInSelection = root.FindTokenOnRightOfPosition(adjustedSpan.Start, includeSkipped: false);
+            var lastTokenInSelection = root.FindTokenOnLeftOfPosition(adjustedSpan.End, includeSkipped: false);
+
+            if (firstTokenInSelection.Kind() == SyntaxKind.None || lastTokenInSelection.Kind() == SyntaxKind.None)
+                return InitialSelectionInfo.Failure(FeaturesResources.Invalid_selection);
+
+            var commonRoot = firstTokenInSelection.GetCommonRoot(lastTokenInSelection);
+            var selectionInExpression = commonRoot is ExpressionSyntax;
+
+            var statusReason = CheckSpan();
+            if (statusReason is not null)
+                return InitialSelectionInfo.Failure(statusReason);
+
+            return CreateInitialSelectionInfo(
+                selectionInExpression, firstTokenInSelection, lastTokenInSelection, cancellationToken);
+
+            string? CheckSpan()
+            {
+                if (firstTokenInSelection.SpanStart > lastTokenInSelection.Span.End)
+                    return FeaturesResources.Selection_does_not_contain_a_valid_token;
+
+                if (!UnderValidContext(firstTokenInSelection) || !UnderValidContext(lastTokenInSelection))
+                    return FeaturesResources.No_valid_selection_to_perform_extraction;
+
+                if (commonRoot == null)
+                    return FeaturesResources.No_common_root_node_for_extraction;
+
+                if (!commonRoot.ContainedInValidType())
+                    return FeaturesResources.Selection_not_contained_inside_a_type;
+
+                if (!selectionInExpression && !commonRoot.UnderValidContext())
+                    return FeaturesResources.No_valid_selection_to_perform_extraction;
+
+                return null;
+            }
+        }
+
+        protected override FinalSelectionInfo UpdateSelectionInfo(
+            InitialSelectionInfo initialSelectionInfo,
+            CancellationToken cancellationToken)
         {
             var root = SemanticDocument.Root;
             var model = SemanticDocument.SemanticModel;
 
             // go through pipe line and calculate information about the user selection
-            selectionInfo = AssignInitialFinalTokens(selectionInfo, firstStatement, lastStatement);
+            var selectionInfo = AssignInitialFinalTokens(initialSelectionInfo);
             selectionInfo = AdjustFinalTokensBasedOnContext(selectionInfo, model, cancellationToken);
-            selectionInfo = AssignFinalSpan(selectionInfo);
-            selectionInfo = ApplySpecialCases(selectionInfo, SemanticDocument.SyntaxTree.Options, _localFunction);
+            selectionInfo = AssignFinalSpan(initialSelectionInfo, selectionInfo);
```

Comment:
>initialSelectionInfo [](http://example.com/codeflow?start=44&length=20)

Seems odd that this and the next line pass in initialSelectionInfo

Reply:
never mind, selectionInfo is also passed in

Reply:
they're doing some work comparing the initial to the final they've computed so far.  (yes, odd).